client/my-sites/checkout/checkout/billing-details-form.jsx

```diff
@@ -0,0 +1,379 @@
+/** @format */
+/**
+ * External dependencies
+ */
+import React, { PureComponent } from 'react';
+import { connect } from 'react-redux';
+import { localize } from 'i18n-calypso';
+import classNames from 'classnames';
+import { debounce, intersection, map, noop } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import config from 'config';
+import wp from 'lib/wp';
+import { isMobile } from 'lib/viewport';
+import { cartItems } from 'lib/cart-values';
+import {
+	addPrivacyToAllDomains,
+	removePrivacyFromAllDomains,
+	setBillingDetails,
+	setDomainDetails,
+	addGoogleAppsRegistrationData,
+} from 'lib/upgrades/actions';
+import QueryBillingContactDetails from 'components/data/query-billing-contact-details';
+import QueryTldValidationSchemas from 'components/data/query-tld-validation-schemas';
+import ContactDetailsFormFields from 'components/domains/contact-details-form-fields';
+import FormButton from 'components/forms/form-button';
+import FormCheckbox from 'components/forms/form-checkbox';
+import InfoPopover from 'components/info-popover';
+import ExtraInfoForm, {
+	tldsWithAdditionalDetailsForms,
+} from 'components/domains/registrant-extra-info';
+import getBillingContactDetails from 'state/selectors/get-billing-contact-details';
+import { updateContactDetailsCache } from 'state/domains/management/actions';
+import { updateBillingContactDetails } from 'state/billing-transactions/actions';
+import { recordTracksEvent } from 'state/analytics/actions';
+import SecurePaymentFormPlaceholder from './secure-payment-form-placeholder';
+import PaymentBox from './payment-box';
+
+const wpcom = wp.undocumented();
+
+export class BillingDetailsForm extends PureComponent {
+	constructor( props ) {
+		super( props );
+		this.debounced = {
+			validateBillingDetails: debounce( this.validateBillingDetails, 300, { leading: true } ),
+			validateDomainDetails: debounce( this.validateDomainDetails, 300, { leading: true } ),
+		};
+		this.state = {
+			showDomainContactDetails: false,
+			billingDetails: null,
+		};
+	}
+
+	componentDidMount() {
+		if ( this.props.recordTracksEvent ) {
+			this.props.recordTracksEvent( 'calypso_checkout_billing_contact_information_view' );
+		}
+	}
+
+	validateBillingDetails = ( fieldValues, onComplete ) => {
+		const validationHandler = ( error, data ) => {
+			const messages = ( data && data.messages ) || {};
+
+			if ( ! error ) {
+				this.setState( { billingDetails: fieldValues } );
+				setBillingDetails( fieldValues );
+
+				if ( ! this.state.showDomainContactDetails ) {
+					setDomainDetails( fieldValues );
+					addGoogleAppsRegistrationData( fieldValues );
```

Comment:
Do we need to call `addGoogleAppsRegistrationData()` on every change for both forms? On the domain details forms we're [calling it on submit only](https://github.com/Automattic/wp-calypso/blob/217d0de/client/my-sites/checkout/checkout/domain-details-form.jsx#L235).

There's a slight regression in that it triggers a change in the cart store, and therefore a change in the cart component (because it's listening out for `hasPendingServerUpdates`)

![update-cart](https://user-images.githubusercontent.com/6458278/47891308-ccfefb00-dea6-11e8-8626-d56b16314ffa.gif)

Reply:
@ramonjd That's a good call. In the previous checkout flow, we have a submit button on the contact form. When users click on the button to move to the next step - the checkout page, we validate the contact information they input before showing the checkout form. But now we don't have the gap though we still need to validate the information. Should we recall the submit button to the contact form?
@sixhours What do you think?

Reply:
Hmm, adding a button in the middle of the page would complicate things and it's not a common pattern. :-/ I'm not sure of a programmatic way around this. Is there a severe performance drawback (or other issue) to calling the function repeatedly? I can see it's not a great for rendering, but adding a button to check the inputs mid-flow basically brings us back to the two-step checkout we're trying to avoid with this A/B test.

Reply:
@sixhours The client-side performance would be fine even if the sidebar and the checkout button keep refreshing every time the values of form fields change. But the server-side can be serious since the number of our users is tremendously huge. I also agree with you on the point that adding the validation button to the form may harm the user experience.